### ISSUE TITLE
ValueError: Amounts of different currencies can't be compared.

### ISSUE THREAD

**Issue by hsoft** [hsoft/moneyguru#442]
By having reconciled entries of different currencies, we end up with a crash like this one:

```
Traceback (most recent call last):
  File "/usr/bin/moneyguru", line 67, in <module>
    sys.exit(main(sys.argv))
  File "/usr/bin/moneyguru", line 55, in main
    mgapp = MoneyGuru()
  File "/usr/share/moneyguru/qt/app.py", line 58, in __init__
    self.doc.open(self.prefs.recentDocuments[0])
  File "/usr/share/moneyguru/qt/controller/document.py", line 70, in open
    self.model.load_from_xml(docpath)
  File "/usr/share/moneyguru/core/document.py", line 1167, in load_from_xml
    loader.load()
  File "/usr/share/moneyguru/core/loader/base.py", line 363, in load
    self.oven.cook(datetime.date.min, until_date=None)
  File "/usr/share/moneyguru/core/model/oven.py", line 153, in cook
    self._cook_splits(account, splits)
  File "/usr/share/moneyguru/core/model/oven.py", line 80, in _cook_splits
    split2reconciledbal = self._cook_reconciliation_balances(splits, entries.balance_of_reconciled())
  File "/usr/share/moneyguru/core/model/oven.py", line 72, in _cook_reconciliation_balances
    balance += split.amount
ValueError: Amounts of different currencies can't be compared.
```

This error has been locked down through moneyGuru versions and it's now, I think, not possible to have it "naively", but I've received reports of getting this error when opening old moneyGuru documents, and I can indeed reproduce the problem:

1. New document
2. New account with two reconciled entries
3. Save
4. Open the document with a text editor
5. Go change the currency of one of the entries
6. Open the file
7. Crash

**Comment by ancientHacker** [comment]
So I have reproduced this on Mac, and played a bit with currency behaviors, and I'm thinking I'd like to take a shot at it.  Let me know if that's OK.

**Comment by ancientHacker** [comment]
Analysis: What's going on here is that, because the UI disables reconciliation for amounts that are not in the account currency, the code that cooks balances for reconciled items doesn't expect to do currency conversion when drawing its balance (in contrast with the code that cooks balances for non-reconciled items).  If you ever end up with a data file that has a reconciled item in a non-account currency, the load of the file will fail (as documented in the issue) because it's summing all of the reconciled amounts without first converting them to the account currency.

Proposed fix: Do currency conversion for reconciliation balances in exactly the same way they are done for regular balances.  This requires passing the account into the reconciliation cooker, which seems harmless.  It has the added advantage of keeping reconciled balances and regular balances consistent.  It might some day allow for marking non-account currency transactions as reconciled (currently this is prevented by the UI), but because currency conversion tables can (in theory) change retroactively this would allow reconciled balances to change retroactively.  That might not be a desirable thing.

Additional: It is also possible to provoke this bug by changing the currency on an account after reconciling some transactions, and then trying to reconcile a new transaction in the (new) default currency.  The UI doesn't currently allow this change on an account that has transactions in it, but I'm not sure that was true with early versions of MoneyGuru.  That might explain why some older data files manifest this bug.  The proposed fix handles this case as well.

**Comment by hsoft** [comment]
Thanks! I'll review this soon.

**Comment by ancientHacker** [comment]
@hsoft and I have discussed this and agree that, while the proposed fix works, it actually conceals the underlying data problem: accounts should not contain reconciled transactions that are in a different currency than the account itself.  (Although MoneyGuru is now quite effective at preventing such transactions from being reconciled, it was more permissive in earlier releases.)  The right fix, then, is to undo the data problems that may have crept into an older (or manually edited) file by _unreconciling_ off-currency transactions when the file is first loaded.

**Comment by ancientHacker** [comment]
Hi @hsoft, I submitted a pull request to fix this a couple of days ago.  Take a look when you have a chance. -d.

**Comment by hsoft** [comment]
Fixed in #448
